### PR TITLE
[Develop] Feature/5 add create step

### PR DESCRIPTION
**Pull request by Te7a-Houdini**
this **PR** resolves part of issue #7 .

what has been **done** so far :

- Ability to generate **ItemCreatePage** using **RelayForm** example component

- Add **Create Link**  to **ItemList**

- Remove hard coded **PostsListEntry** and **postInfo** from **ItemsList**

- Ask the user about **itemCreateEntryFields** when using plop to generate template

- Modify eslint rules to allow  writing strings in **Link component** like **< link to="posts/create" />**

what's **remaining** : 

- Allow nested inputs names in [Relay Form Fields](https://github.com/Te7a-Houdini/bractal/blob/1b404ac9e77ea3b654ca7bb83a624868c2bbc8d8/plop/templates/module_generator_master_details/sampleModule/containers/ItemCreatePage.js.hbr#L22) as [current implementation in TcombHelpers.js](https://github.com/BadrIT/bractal/blob/develop/src/modules/coreUI/components/forms/TcombHelpers.js#L126) doesn't allow to return a **nested objects** for example 

in **createPost mutation** the object that needs to be send for GraphQl Server must be like this

```js
data:{
     title: "Nice Article33 !" 
     description: "And thi44s is it's description"
     author: {
       connect: {
         id: "cjljjqw1e3i3c0b66w5vphamg"
       }
     }
   }
```
and current **RelayFormFields** only allow one level of input naming [like in sample](https://github.com/BadrIT/bractal/blob/master/docs/assets/relay_form/sample.relayForm.js#L123)

- Implementation of **onSucess** method

- Update the Project **ReadMe** with the new **ItemCreatePage** section

### CODE REVIEW

File: plop/templates/module_generator_master_details/sampleModule/containers/ItemCreatePage.js.hbr [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import { BasicButton } from '~/modules/coreUI/components/basic/Button';
+
+const FormMutation = graphql`
+  mutation {{uiItemName}}CreatePageMutation (
+   $data : {{uiItemName}}CreateInput!
+  ) {
+    create{{uiItemName}}(
+     data : $data
+    ) {
+        {{itemCreateEntryFields}}
```
Comment (mostafaelganainy) [inline plop/templates/module_generator_master_details/sampleModule/containers/ItemCreatePage.js.hbr]: Adjust the indentation

File: plop/templates/module_generator_master_details/sampleModule/containers/ItemCreatePage.js.hbr [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import { XXXXLargeSpacer, LargeSpacer } from '~/modules/coreUI/components/layouts/helpers/Spacers';
+import { ErrorLabel } from '~/modules/coreUI/components/basic/Labels';
+import { BasicButton } from '~/modules/coreUI/components/basic/Button';
+
+const FormMutation = graphql`
+  mutation {{uiItemName}}CreatePageMutation (
+   $data : {{uiItemName}}CreateInput!
+  ) {
+    create{{uiItemName}}(
+     data : $data
```
Comment (mostafaelganainy) [inline plop/templates/module_generator_master_details/sampleModule/containers/ItemCreatePage.js.hbr]: Adjust the indentation

**Comment by Te7a-Houdini** [comment]
@mostafaelganainy  i wanted to know if allowing nested form objects issue i mentioned have been fixed or not ?

as the main issue am trying to solve here didn't mention that RelayForm won't allow nested objects

File: src/modules/coreUI/components/forms/RelayForm.jsx [hunk 3]
```diff
@@ -80,6 +80,19 @@ class RelayForm extends Component {
 
   getValue = () => this.Form.getValue();
 
+  getCommitFormMutationVariables = () => {
```
[the remark this answers is not in the document]
Comment (Te7a-Houdini) [reply src/modules/coreUI/components/forms/RelayForm.jsx]: i got your point , but actually , the solution you mentioned won't work cause it requires to **wrap the whole object into data** and will produce this error , the reason for the error that we are
[using object spreading without wrapping the variables constant under data](https://github.com/BadrIT/bractal/blob/develop/src/modules/coreUI/components/forms/RelayForm.jsx#L20) .
so i was into two decisions wither i modify only the variables under data which is the simplest solution , or i extract the logic into a function that can be overridden .

 so i choose the other approach to allow more flexibility  , but  i think in this case of post , we can only make the data as a wrapper and it will fix the issue .

```javascript
{
    data : {
          title : 'title value',
          description : 'description value',
           author : { 
                   connect : { 
                        id : "cjljjqw1e3i3c0b66w5vphamg"
                  }
         }
    }
}

const variables = { 
  data : {
      {
         ...formValues,
          ...addiontalMutationVariables,
    }
   }
}
```
 so to summarize we need the data key to be a wrapper , and also i wanted to point another issue i thought about it , what if for example we want to have slug from title, (i know that it can be handled from backend) but i am pointing to the issue of a **computed field** that is based on another field , (the solution of additionalMutationVariables won't work , cause we have **no access on Form object** so we can't make something like this **slug(form.getComponent('title').props.value)** )  .

```javascript
{ 
   data : {
        title : 'New Post' ,
        slug : 'new-post'      
   }
}
```
i hope i cleared why i choose extracting the logic to a function , and also i want to know your thoughts about it .

screenshot from 2018-09-12 10-44-16

File: plop/templates/module_generator_master_details/sampleModule/containers/ItemCreatePage.js.hbr [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      create{{uiItemName}} (
+        data : $data
+      ) {
+          {{itemCreateEntryFields}}
+      }
+  }
+`;
+
+const RelayFormFields = {
+  fields: [],
```
[the remark this answers is not in the document]
Comment (Te7a-Houdini) [reply plop/templates/module_generator_master_details/sampleModule/containers/ItemCreatePage.js.hbr]: it's left blank as a convention for any new Create template , and the developer will fill it .

File: src/modules/coreUI/components/forms/RelayForm.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mostafaelganainy) [reply src/modules/coreUI/components/forms/RelayForm.jsx]: Okay, now I got why you were confused. But extraction to a method, isn't the way to go about it. It's all about how you write the mutation. Check this :

```javascript
 mutation CreatePost (
    $title: String!,
    $description: String!,
    $author: SOME_GRAPHQL_TYPE!,
 ) {
    forgot_password_verify_code (
      data : {
         title: $title,
         description: $description,
         author: $author,
    ) {      
      .......
    }
  }
```

And it would be consumed the same as I've illustrated above.

[the remark this answers is not in the document]
Comment (Te7a-Houdini) [reply src/modules/coreUI/components/forms/RelayForm.jsx]: @mostafaelganainy  aha ok got your point

**Comment by mostafaelganainy** [comment]
@Te7a-Houdini : I'll continue the work from where your left. I assume that you're fine with that.

**Comment by Te7a-Houdini** [comment]
@mostafaelganainy yes sure go ahead
